File: CS/Q351285/Form1.cs
using System;
using System.IO;
using System.Windows.Forms;
using DevExpress.XtraTreeList;
using System.Collections.Generic;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraEditors;

namespace Q351285 {
    public partial class Form1 : XtraForm {
        private string treeListDataFile = "data.xml";
        public Form1() {
            InitializeComponent();
        }
        private void Form1_Load(object sender, EventArgs e) {
            InitData();
            InitTreeList();
        }
        private void InitTreeList() {
            treeList1.ForceInitialize();
            UpdateNodesPositions(treeList1.Nodes);
            treeList1.ExpandAll();
        }
        private void InitData() {
            if(File.Exists(treeListDataFile)) {
                dataTable1.ReadXml(treeListDataFile);
            } else {
                FillTable();
            }
        }
        private void FillTable() {
            dataTable1.Rows.Add(1, 0, "A", 0/*initially, this value is used to determine the node position*/);
            dataTable1.Rows.Add(2, 1, "B", 1);
            dataTable1.Rows.Add(3, 1, "C", 2);
            dataTable1.Rows.Add(4, 0, "D", 3);
            dataTable1.Rows.Add(5, 4, "E", 4);
        }
        private void Form1_FormClosing(object sender, FormClosingEventArgs e) {
            SaveData();
        }
        private void SaveData() {
            dataTable1.WriteXml(treeListDataFile);
        }
        private void UpdateNodesPositions(TreeListNodes nodes) {
            var ns = new List<TreeListNode>();
            foreach (TreeListNode n in nodes) {
                ns.Add(n);
            }
            foreach (TreeListNode n in ns) {
                UpdateNodesPositions(n.Nodes);
                n.TreeList.SetNodeIndex(n, Convert.ToInt32(n.GetValue("Order")));
            }
        }
        private void treeList1_AfterDragNode(object sender, AfterDragNodeEventArgs e) {
            SaveNewRecordPosition(e);
        }
        private void SaveNewRecordPosition(NodeEventArgs e) {
            var nodes = e.Node.ParentNode == null ? e.Node.TreeList.Nodes
                        : e.Node.ParentNode.Nodes;
            for(var i = 0; i < nodes.Count; i++) {
                nodes[i].SetValue(colSort, i);
            }
        }
    }
}

File: VB/Q351285/Form1.vb
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList
Imports System.Collections.Generic
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraEditors

Namespace Q351285

    Public Partial Class Form1
        Inherits XtraForm

        Private treeListDataFile As String = "data.xml"

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            InitData()
            InitTreeList()
        End Sub

        Private Sub InitTreeList()
            treeList1.ForceInitialize()
            UpdateNodesPositions(treeList1.Nodes)
            treeList1.ExpandAll()
        End Sub

        Private Sub InitData()
            If File.Exists(treeListDataFile) Then
                dataTable1.ReadXml(treeListDataFile)
            Else
                FillTable()
            End If
        End Sub

        Private Sub FillTable()
            dataTable1.Rows.Add(1, 0, "A", 0)
            dataTable1.Rows.Add(2, 1, "B", 1)
            dataTable1.Rows.Add(3, 1, "C", 2)
            dataTable1.Rows.Add(4, 0, "D", 3)
            dataTable1.Rows.Add(5, 4, "E", 4)
        End Sub

        Private Sub Form1_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs)
            SaveData()
        End Sub

        Private Sub SaveData()
            dataTable1.WriteXml(treeListDataFile)
        End Sub

        Private Sub UpdateNodesPositions(ByVal nodes As TreeListNodes)
            Dim ns = New List(Of TreeListNode)()
            For Each n As TreeListNode In nodes
                ns.Add(n)
            Next

            For Each n As TreeListNode In ns
                UpdateNodesPositions(n.Nodes)
                n.TreeList.SetNodeIndex(n, Convert.ToInt32(n.GetValue("Order")))
            Next
        End Sub

        Private Sub treeList1_AfterDragNode(ByVal sender As Object, ByVal e As NodeEventArgs)
            SaveNewRecordPosition(e)
        End Sub

        Private Sub SaveNewRecordPosition(ByVal e As NodeEventArgs)
            Dim nodes = If(e.Node.ParentNode Is Nothing, e.Node.TreeList.Nodes, e.Node.ParentNode.Nodes)
            For i = 0 To nodes.Count - 1
                nodes(i).SetValue(colSort, i)
            Next
        End Sub
    End Class
End Namespace
